src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -102,13 +111,228 @@ private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(Document
+
+                    startLine += startTokens[i];
+                    startCharacter += startTokens[i + 1];
+                }
+
+                // Compute the absolute beginning line/character of the ending tokens.
+                // This loop also determines which tokens are duplicates.
+                var endLine = 0;
+                var endCharacter = 0;
+                var tokensToSkip = 0;
+                for (var i = 0; i < endTokens.Length; i += 5)
+                {
+                    if (endTokens[i] != 0)
+                    {
+                        endCharacter = 0;
+                    }
+
+                    endLine += endTokens[i];
+                    endCharacter += endTokens[i + 1];
+
+                    if (endLine > startLine)
+                    {
+                        break;
+                    }
+                    else if (endLine < startLine || endCharacter <= startCharacter)
+                    {
+                        tokensToSkip += 5;
+                        continue;
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+
+                // Skip any duplicate tokens
+                if (tokensToSkip > 0)
+                {
+                    endTokens = endTokens.Skip(tokensToSkip).ToArray();
+                }
+
+                // The ending set of tokens will need to have its starting line modified to be relative
+                // to the end of the starting set of tokens.
+                if (endLine > startLine)
+                {
+                    endTokens[0] = endLine - startLine;
+                }
+
+                var combinedTokens = startTokens.Concat(endTokens);
+                var semanticTokens = new SemanticTokens { Data = combinedTokens.ToImmutableArray() };
+
+                if (isCSharpFinalized)
+                {
+                    var combinedRange = computeStart ? new Range(range.Start, bestMatch.Range.End) : new Range(bestMatch.Range.Start, range.End);
+                    CacheTokens(textDocumentIdentifier.Uri, semanticVersion, combinedRange, semanticTokens);
+                }
+
+                return semanticTokens;
+            }
+        }
+
+        private void CacheTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range range,
+            SemanticTokens tokens)
+        {
+            if (!_cachedResponses.TryGetValue(uri, out var semanticVersionCache))
+            {
+                semanticVersionCache = new MemoryCache<VersionStamp, MemoryCache<Range, SemanticTokens>>(sizeLimit: MaxSemanticVersionsPerDoc);
+                _cachedResponses.Set(uri, semanticVersionCache);
+            }
+
+            if (!semanticVersionCache.TryGetValue(semanticVersion, out var tokensCache))
+            {
+                tokensCache = new MemoryCache<Range, SemanticTokens>(sizeLimit: MaxRangesPerSemanticVersion);
```

Comment:
We should probably do a follow up here to optimize how we cache ranges in the document so we're not holding onto loads of overlapping bits + could potentially enable us to cache ranges in groups in some arbitrary group of lines or something